### PR TITLE
Make the RAND code available from within the FIPS module

### PR DESCRIPTION
**Pull request by mattcaswell**
This enables calls to RAND_*() functions from within the FIPS module. This is required in order to bring in support for BIGNUM, and the various asymmetric algorithms.

I've added a dummy call into the FIPS module just to demonstrate that it works. This will be removed in due course when we no longer need it.

### CODE REVIEW

File: crypto/async/async.c [hunk 1]
```diff
@@ -374,7 +376,8 @@ int ASYNC_init_thread(size_t max_size, size_t init_size)
     return 0;
 }
 
-void async_delete_thread_state(void)
+/* OPENSSL_CTX ignored for now */
```
Comment (slontis) [inline crypto/async/async.c]: TODO 3.0 ?

File: crypto/context.c [hunk 2]
```diff
@@ -155,11 +169,7 @@ static int openssl_ctx_init_index(OPENSSL_CTX *ctx, int static_index,
     int idx;
 
 #ifndef FIPS_MODE
-    if (ctx == NULL) {
-        if (!RUN_ONCE(&default_context_init, do_default_context_init))
-            return 0;
-        ctx = default_context;
-    }
+    ctx = openssl_ctx_get_concrete(ctx);
```
Comment (slontis) [inline crypto/context.c]: Do we want this kind of pattern? Or would it be better to just call the function always and 
#ifdef FIPS_MODE
# define openssl_ctx_get_concrete(ctx) ctx
#endif

File: crypto/err/err.c [hunk 3]
```diff
@@ -685,7 +685,8 @@ const char *ERR_reason_error_string(unsigned long e)
     return ((p == NULL) ? NULL : p->string);
 }
 
-void err_delete_thread_state(void)
+/* OPENSSL_CTX ignored for now */
```
Comment (slontis) [inline crypto/err/err.c]: TODO 3.0

File: crypto/include/internal/cryptlib_int.h [hunk 4]
```diff
@@ -7,17 +7,16 @@
...
-    int err_state;
-    int rand;
-};
-
-int ossl_init_thread_start(uint64_t opts);
+int ossl_init_thread_start(void *arg,
+                           OSSL_thread_stop_handler_fn handfn);
+int init_thread(void);
+void cleanup_thread(void);
+void ossl_ctx_thread_stop(void *arg);
```
Comment (slontis) [inline crypto/include/internal/cryptlib_int.h]: These start/stop names look different

File: crypto/initthread.c [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+} destructor_key = { -1 };
+
+static void ossl_init_thread_destructor(void *hands)
+{
+    ossl_init_thread_stop(NULL, (THREAD_EVENT_HANDLER **)hands);
+}
+
+int init_thread(void)
+{
+
```
Comment (slontis) [inline crypto/initthread.c]: NIT: blankline

File: crypto/initthread.c [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        curr->handfn(curr->arg);
+        prev = curr;
+        curr = curr->next;
+        if (prev == *hands)
+            *hands = NULL;
+        OPENSSL_free(prev);
+    }
+
+    if (freehands)
+        OPENSSL_free(hands);
```
Comment (slontis) [inline crypto/initthread.c]: Should *hands be NULL here?

File: crypto/initthread.c [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        if (arg != NULL && curr->arg != arg) {
+            curr = curr->next;
+            freehands = 0;
+            continue;
+        }
+        curr->handfn(curr->arg);
+        prev = curr;
+        curr = curr->next;
+        if (prev == *hands)
+            *hands = NULL;
```
Comment (slontis) [inline crypto/initthread.c]: Doesnt this *hands get returned?

File: crypto/initthread.c [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+int ossl_init_thread_start(void *arg, OSSL_thread_stop_handler_fn handfn)
+{
+    THREAD_EVENT_HANDLER **hands;
+    THREAD_EVENT_HANDLER *hand;
+#ifdef FIPS_MODE
+    OPENSSL_CTX *ctx = arg;
+
+    /*
+     * In FIPS mode the list of THREAD_EVENT_HANDLERs unique per combination of
```
Comment (slontis) [inline crypto/initthread.c]: NIT is unique

File: crypto/initthread.c [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+     * OPENSSL_CTX and thread. This is because in FIPS mode each OPENSSL_CTX
+     * gets informed about thread stop events individually.
+     */
+    CRYPTO_THREAD_LOCAL *local
+        = openssl_ctx_get_data(ctx, OPENSSL_CTX_THREAD_EVENT_HANDLER_INDEX,
+                               &thread_event_ossl_ctx_method);
+#else
+    /*
+     * Outside of FIPS mode the list of THREAD_EVENT_HANDLERs is unque per
```
Comment (slontis) [inline crypto/initthread.c]: NIT: unique

File: crypto/rand/drbg_lib.c [hunk 10]
```diff
@@ -141,6 +147,161 @@ static int is_digest(int type)
...
+#ifndef FIPS_MODE
+    /*
+     * We need to ensure that base libcrypto thread handling has been
+     * initialised.
+     */
+     OPENSSL_init_crypto(0, NULL);
+#endif
+
+    if (!CRYPTO_THREAD_init_local(&dgbl->private_drbg, NULL))
+        goto err1;
```
Comment (slontis) [inline crypto/rand/drbg_lib.c]: I assume all these goto's mean that CRYPTO_THREAD_cleanup_local() doesnt handle passing NULLS - looks  messy

**Comment by slontis** [comment]
I havent gone through the whole thing yet - but it seems that a large percentage of this PR is related to threads. Maybe the rand changes should be in a seperate PR?

**Comment by slontis** [comment]
travis errors are valid (except for the timeout ones)

File: crypto/include/internal/cryptlib_int.h (hunk 4 above)
Comment (mattcaswell) [reply crypto/include/internal/cryptlib_int.h]: Not sure I understand your point. There isn't a "start" equivalent of `ossl_ctx_thread_stop`

File: crypto/initthread.c (hunk 7 above)
Comment (mattcaswell) [reply crypto/initthread.c]: I refactored this to make things a bit clearer.

File: crypto/rand/drbg_lib.c (hunk 10 above)
Comment (mattcaswell) [reply crypto/rand/drbg_lib.c]: Well, `dgbl->private_drbg` is not NULL if there's an error, just uninitialised - so we shouldn't attempt to use it in anyway.

**Comment by mattcaswell** [comment]
I've attempt to respond to the comments so far and have also tried to fix the various travis issues. I've also rebased and force pushed.

I'm going to take a look to see if I can easily split this up into smaller PRs to make it easier to review.

**Comment by mattcaswell** [comment]
I've split out various bits of this PR into other PRs. See #9038, #9039, and #9040. Once all of those are merged this PR will need to be rebased to pick up the last few commits to complete the move of the RAND code into the FIPS module. In the meantime I've set this PR to be WIP.

File: crypto/rand/drbg_lib.c (hunk 10 above)
Comment (richsalz) [reply crypto/rand/drbg_lib.c]: Since dgbl is allocated via zalloc, and since NULL is all-bits-zero (via the sanity test), it is initialized.

Comment (mattcaswell) [reply crypto/rand/drbg_lib.c]: It's the `private_drbg` field we are specifically concerned with there. And yes it will get set to 0 on create. However, when I said it was unitialised, I meant it has not been initialised via a call to `CRYPTO_THREAD_init_local`. We should not be calling `CRYPTO_THREAD_cleanup_local` on keys that have not been initialised in that way. We cannot know how `CRYPTO_THREAD_cleanup_local` would treat a key that is set to 0. It most likely would be fine but we cannot say for sure. For example in a pthreads implementation `CRYPTO_THREAD_cleanup_local` calls `pthread_key_delete` which has this description: "The pthread_key_delete() function shall delete a thread-specific data key previously returned by pthread_key_create()". It says nothing about what happens if you pass it zero data that was not created by `pthread_key_create` - so at best this behaviour is undefined.

**Comment by mattcaswell** [comment]
I rebased this now that #9038 has gone in, and to include the latest changes from #9039.

File: crypto/include/internal/cryptlib_int.h (hunk 4 above)
Comment (levitte) [reply crypto/include/internal/cryptlib_int.h]: What about `ossl_init_thread_start`?

Comment (mattcaswell) [reply crypto/include/internal/cryptlib_int.h]: The `stop` equivalent of `ossl_init_thread_start` is `ossl_init_thread_stop`. There is no "start" equivalent of `ossl_ctx_thread_stop`.

**Comment by mattcaswell** [comment]
Needs #9186 before this can progress.

File: doc/man3/OPENSSL_CTX.pod [hunk 11]
```diff
@@ -20,7 +20,10 @@ Applications may allocate their own, but may also use C<NULL> to use
 the internal default context with functions that take a C<OPENSSL_CTX>
 argument.
 
-OPENSSL_CTX_new() creates a new OpenSSL library context.
+OPENSSL_CTX_new() creates a new OpenSSL library context. When a non default
```
Comment (levitte) [inline doc/man3/OPENSSL_CTX.pod]: Line break after sentence end, please

File: doc/man3/RAND_DRBG_get0_master.pod [hunk 12]
```diff
@@ -21,26 +27,35 @@ RAND_DRBG_get0_private
...
-RAND_DRBG_get0_public() returns a pointer to the <public> DRBG instance.
+OPENSSL_CTX_get0_public_drbg() returns a pointer to the I<public> DRBG instance
+for the given OPENSSL_CTX B<ctx>.
 
-RAND_DRBG_get0_private() returns a pointer to the <private> DRBG instance.
+OPENSSL_CTX_get0_private_drbg() returns a pointer to the I<private> DRBG instance
+for the given OPENSSL_CTX B<ctx>.
 
+In all the above cases the B<ctx> parameter can
+be NULL in which case the default OPENSSL_CTX is used. RAND_DRBG_get0_master(),
```
Comment (levitte) [inline doc/man3/RAND_DRBG_get0_master.pod]: Line break after sentence end, please

**Comment by levitte** [comment]
Please rebase then ping, I'll try to review quickly.

**Comment by mattcaswell** [comment]
I've rebased this PR. Note the feedback comments above were for commits no longer in this PR. Note this PR now also includes the commits from #9186, so that needs to go in first before I can take this out of WIP.

**Comment by mattcaswell** [comment]
Finally all the dependent PRs are in and I am able to take this out of WIP!

I've rebased on the latest master. Please review!

File: crypto/hmac/build.info [hunk 13]
```diff
@@ -1,3 +1,6 @@
 LIBS=../../libcrypto
 SOURCE[../../libcrypto]=\
         hmac.c hm_ameth.c hm_meth.c
+
+SOURCE[../../providers/fips]=\
+        hmac.c hm_meth.c
```
Comment (levitte) [inline crypto/hmac/build.info]: May I suggest this?

```
$COMMON=hmac.c hm_meth.c

LIBS=../../libcrypto
SOURCE[../../libcrypto]=$COMMON hm_ameth.c
SOURCE[../../providers/fips]=$COMMON
```

File: crypto/rand/build.info [hunk 14]
```diff
@@ -3,3 +3,8 @@ SOURCE[../../libcrypto]=\
         randfile.c rand_lib.c rand_err.c rand_crng_test.c rand_egd.c \
         rand_win.c rand_unix.c rand_vms.c drbg_lib.c drbg_ctr.c rand_vxworks.c \
         drbg_hash.c drbg_hmac.c
+
+# FIPS Module
+SOURCE[../../providers/fips]=\
+        rand_lib.c rand_crng_test.c rand_win.c rand_unix.c  rand_vms.c \
+        drbg_lib.c drbg_ctr.c rand_vxworks.c drbg_hash.c drbg_hmac.c
```
Comment (levitte) [inline crypto/rand/build.info]: SImilarly:

```
$COMMON=rand_lib.c rand_crng_test.c rand_win.c rand_unix.c  rand_vms.c \
        drbg_lib.c drbg_ctr.c rand_vxworks.c drbg_hash.c drbg_hmac.c

LIBS=../../libcrypto
SOURCE[../../libcrypto]=$COMMON randfile.c rand_err.c rand_egd.c
SOURCE[../../providers/fips]=$COMMON
```

File: crypto/hmac/hm_meth.c [hunk 15]
```diff
@@ -152,13 +152,16 @@ static int hmac_ctrl_str(EVP_MAC_IMPL *hctx, const char *type,
...
+#ifndef FIPS_MODE
+    /* Not supported in FIPS mode due to the implict fetch of the md */
     if (strcmp(type, "digest") == 0) {
         const EVP_MD *d = EVP_get_digestbyname(value);
 
         if (d == NULL)
             return 0;
         return hmac_ctrl_int(hctx, EVP_MAC_CTRL_SET_MD, d);
     }
+#endif
```
Comment (levitte) [inline crypto/hmac/hm_meth.c]: This is problematic, 'cause directly fetched MDs are broken for the moment, since sizes and flags haven't been transferred.  I need to get back to the fixing PR, thanks for the reminder...

**Comment by mattcaswell** [comment]
Hmm. This is strange. Travis is reporting a doc-nits failure:

````
*** ERROR: Apparent command =head1 not preceded by blank line at line 8 in file doc/internal/man3/ossl_init_thread_deregister.pod

make: *** [doc-nits] Error 1

+// MAKE DOC-NITS FAILED
````

However I don't get the above error locally (doc-nits passes)!!? However if I add a blank line at line 8 as suggested by the error message I get this locally:

````
*** WARNING: empty section in previous paragraph at line 9 in file doc/internal/man3/ossl_init_thread_deregister.pod
Makefile:524: recipe for target 'doc-nits' failed
make: *** [doc-nits] Error 1
````
In any case, neither of these issues is directly related to this PR.

File: crypto/hmac/build.info (hunk 13 above)
Comment (mattcaswell) [reply crypto/hmac/build.info]: Ah, yes. This code was written before you introduced that capability. Fixed.

File: crypto/rand/build.info (hunk 14 above)
Comment (mattcaswell) [reply crypto/rand/build.info]: Fixed.

**Comment by mattcaswell** [comment]
Fixup commit pushed addressing the feedback, and also fixing a windows linking issue. I've not addressed the doc-nits issue because that is (presumably) also a problem on master so probably is better done in a different PR.

**Comment by mattcaswell** [comment]
doc-nits issue fixed in #9192

**Comment by levitte** [comment]
Looks good to me. I would like to know what @mspncp has to say, though.

**Comment by mspncp** [comment]
> Looks good to me. I would like to know what @mspncp has to say, though.

I started taking a look, but it might take a while to complete it, since I'm currently on vacation.

**Comment by levitte** [comment]
Oh, sorry, forgot about that...

File: crypto/rand/drbg_hash.c [hunk 16]
```diff
@@ -303,23 +304,39 @@ static RAND_DRBG_METHOD drbg_hash_meth = {
...
-    const EVP_MD *md;
+    EVP_MD *md;
     RAND_DRBG_HASH *hash = &drbg->data.hash;
 
+#ifndef FIPS_MODE
     /* Any approved digest is allowed */
-    md = EVP_get_digestbynid(drbg->type);
+    md = EVP_MD_meth_dup(EVP_get_digestbynid(drbg->type));
+#else
+    /* TODO(3.0): Fill this out with the complete list of allowed digests */
```
Comment (mspncp) [inline crypto/rand/drbg_hash.c]: Continuing my previous comment: For the moment, I would add all algorithms available for the default provider and change the TODO comment to remove the non-fips algorithms from the switch.

File: crypto/rand/drbg_hmac.c [hunk 17]
```diff
@@ -197,23 +198,37 @@ static RAND_DRBG_METHOD drbg_hmac_meth = {
...
 int drbg_hmac_init(RAND_DRBG *drbg)
 {
-    const EVP_MD *md = NULL;
+    EVP_MD *md = NULL;
     RAND_DRBG_HMAC *hmac = &drbg->data.hmac;
 
+#ifndef FIPS_MODE
     /* Any approved digest is allowed - assume we pass digest (not NID_hmac*) */
-    md = EVP_get_digestbynid(drbg->type);
+    md = EVP_MD_meth_dup(EVP_get_digestbynid(drbg->type));
```
Comment (mspncp) [inline crypto/rand/drbg_hmac.c]: No generic method lookup for the HMAC-DRBG either. See my comments for the HASH-DRBG.

File: crypto/rand/drbg_hmac.c [hunk 18]
```diff
@@ -197,23 +198,37 @@ static RAND_DRBG_METHOD drbg_hmac_meth = {
...
+#else
+    /* TODO(3.0): Fill this out with the complete list of allowed digests */
```
Comment (mspncp) [inline crypto/rand/drbg_hmac.c]: See my comments for the HASH-DRBG.

File: crypto/hmac/hm_meth.c [hunk 19]
```diff
@@ -152,13 +152,16 @@ static int hmac_ctrl_str(EVP_MAC_IMPL *hctx, const char *type,
 {
     if (!value)
         return 0;
+#ifndef FIPS_MODE
+    /* Not supported in FIPS mode due to the implict fetch of the md */
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply crypto/hmac/hm_meth.c]: "Explicit" fetch occurs when we look up an algorithm using one of the "fetch" functions such as EVP_MD_fetch(). This finds the algorithm and associates it with an implementation from a provider. Conversely "Implicit" fetch is when we don't do that, and use a function such as "EVP_sha256()" which has no provider implementation associated with it. With an algorithm obtained in that way the "fetch" happens automatically when you try to use it - hence "implicit fetch". An important difference between these two types of EVP_MD objects is that explicitly fetched versions are dynamically allocated and must be "freed", whereas implicitly fetched ones do not. That fact makes this code more complicated if we want to convert it to use explicit fetch. Since we are unlikely to need this ctrl in FIPS_MODE I didn't bother.

I replaced this comment with a longer one to try and explain this better.

File: crypto/rand/drbg_hash.c [hunk 20]
```diff
@@ -303,23 +304,39 @@ static RAND_DRBG_METHOD drbg_hash_meth = {
...
 int drbg_hash_init(RAND_DRBG *drbg)
 {
-    const EVP_MD *md;
+    EVP_MD *md;
     RAND_DRBG_HASH *hash = &drbg->data.hash;
 
+#ifndef FIPS_MODE
     /* Any approved digest is allowed */
-    md = EVP_get_digestbynid(drbg->type);
+    md = EVP_MD_meth_dup(EVP_get_digestbynid(drbg->type));
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply crypto/rand/drbg_hash.c]: > I don't think the digest algorithm of the HASH-DRBG should be selectable via generic EVP lookup accross multiple providers
...
> The aes-ctr algorithm is always fetched unconditionally from the provider context.

I think there is some confusion here. Ignoring the FIPS provider for a moment (which is a special case), the DRBG code exists in libcrypto and is not associated with a particular provider (yet). We probably will want to make it possible to "fetch" a DRBG implementation from a provider and use that...but we're not there yet. For now its just in libcrypto. The look up across "multiple providers" will happen anyway when used, even with aes-ctr. It just looks for an implementation of (for example) "AES-128-ECB" and uses it. The default provider (which might not even be available for all we know) may or may not be used.

Now, as I said, the FIPS provider is a special case. It does not link against libcrypto and so can't use the built-in DRBG implementation. Therefore, for that provider only, we are additionally making the DRBG code available inside the FIPS provider. For that provider only, when we do a "fetch" it will only ever find algorithms provided by itself.

Anyway, aside from that confusion, your suggestion is fine and I have refactored things accordingly.

File: crypto/rand/drbg_hash.c (hunk 16 above)
Comment (mattcaswell) [reply crypto/rand/drbg_hash.c]: Done

File: crypto/rand/drbg_hmac.c (hunk 17 above)
Comment (mattcaswell) [reply crypto/rand/drbg_hmac.c]: Done

File: crypto/rand/drbg_hmac.c (hunk 18 above)
Comment (mattcaswell) [reply crypto/rand/drbg_hmac.c]: Done

**Comment by mattcaswell** [comment]
New commits addressing the feedback pushed.

File: crypto/rand/drbg_hash.c [hunk 21]
```diff
@@ -303,23 +305,55 @@ static RAND_DRBG_METHOD drbg_hash_meth = {
...
+    case NID_sha512_224:
+    case NID_sha512_256:
+    case NID_sha3_224:
+    case NID_sha3_256:
+    case NID_sha3_384:
+    case NID_sha3_512:
+#ifndef FIPS_MODE
+    case NID_blake2b512:
+    case NID_blake2s256:
+    case NID_sm3:
```
Comment (mspncp) [inline crypto/rand/drbg_hash.c]: This is more a question to @paulidale and @slontis than a review comment:  [NIST SP 800-90Ar1](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf) does not mention SHA3, BLAKE and SM3. Is it 'legal' nevertheless to use them?

File: crypto/rand/drbg_hmac.c [hunk 22]
```diff
@@ -197,23 +199,53 @@ static RAND_DRBG_METHOD drbg_hmac_meth = {
...
+    case NID_sha512_224:
+    case NID_sha512_256:
+    case NID_sha3_224:
+    case NID_sha3_256:
+    case NID_sha3_384:
+    case NID_sha3_512:
+#ifndef FIPS_MODE
+    case NID_blake2b512:
+    case NID_blake2s256:
+    case NID_sm3:
```
Comment (mspncp) [inline crypto/rand/drbg_hmac.c]: Same question to @paulidale and @slontis.

File: crypto/rand/rand_crng_test.c [hunk 23]
```diff
@@ -64,27 +65,29 @@ static const OPENSSL_CTX_METHOD rand_crng_ossl_ctx_method = {
 };
 
 int rand_crngt_get_entropy_cb(OPENSSL_CTX *ctx,
+                              RAND_POOL *pool,
```
Comment (mspncp) [inline crypto/rand/rand_crng_test.c]: What is the rationale behind the changes in `rand_crng_test.c`?

File: crypto/rand/rand_win.c [hunk 24]
```diff
@@ -162,7 +162,7 @@ int rand_pool_add_additional_data(RAND_POOL *pool)
     return rand_pool_add(pool, (unsigned char *)&data, sizeof(data), 0);
 }
 
-# if !OPENSSL_API_1_1_0
+# if !OPENSSL_API_1_1_0 || defined(FIPS_MODE)
```
Comment (mspncp) [inline crypto/rand/rand_win.c]: Are you sure to enable it in FIPS mode? Or did you mean
```
# if !OPENSSL_API_1_1_0 && !defined(FIPS_MODE)
```

File: crypto/rand/rand_crng_test.c (hunk 23 above)
Comment (mspncp) [reply crypto/rand/rand_crng_test.c]: Is it a fixup for 4e297b7441a0 which you made along the way?

**Comment by mattcaswell** [comment]
> BTW: Is it actually possible by now to run and check the DRBG code of the FIPS provider in the debugger (on linux)? If yes, what do I need to set it up?

With this PR, yes. Set a breakpoint on line 120 of `providers/fips/fipsprov.c` and then run `evp_extra_test`:

````
$ OPENSSL_MODULES=providers util/shlib_wrap.sh gdb --args test/evp_extra_test -test 13
GNU gdb (Ubuntu 8.1-0ubuntu3) 8.1.0.20180409-git
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from test/evp_extra_test...done.
(gdb) break fipsprov.c:120
No source file named fipsprov.c.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (fipsprov.c:120) pending.
(gdb) run
Starting program: /home/matt/dev/openssl-write/test/evp_extra_test -test 13
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
    # Subtest: test_EVP_MD_fetch
    1..5
    ok 1 - iteration 1
    ok 2 - iteration 2
    ok 3 - iteration 3

Breakpoint 1, dummy_evp_call (provctx=0x5555557c0530)
    at providers/fips/fipsprov.c:120
120	    if (RAND_DRBG_bytes(drbg, randbuf, sizeof(randbuf)) <= 0)
````

**Comment by mspncp** [comment]
Thanks for the instructions. I'll give it a try this evening...

File: crypto/rand/drbg_hash.c (hunk 21 above)
Comment (paulidale) [reply crypto/rand/drbg_hash.c]: Section 10.1 opens by saying that the hash DRBG mechanisms _have been designed to use any approved hash function_.  Based on this SHA3 **would** be legal.

BLAKE and SM3 cannot be FIPS validated.

File: crypto/rand/drbg_hmac.c (hunk 22 above)
Comment (paulidale) [reply crypto/rand/drbg_hmac.c]: I think it has the same answer.  10.1.2 again mentions an _approved_ hash function.

File: crypto/rand/drbg_hash.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/rand/drbg_hash.c]: I'd lean towards making the digest selection completely open.  It provides a degree of flexibility in case something untoward happens to one of the digests.  However, I'm happy enough with the current switch statement to not be too bothered (there are plenty of choices available and it avoids an explicit XOF check).

One could argue that the CTR-DRBG should have this flexibility too (ARIA and Camellia should both drop in instead of AES).  Still, for some point in the future rather than now.

[the remark this answers is not in the document]
Comment (mspncp) [reply crypto/rand/drbg_hash.c]: I don't mind, as long as the implementation is consistent for the three DRBG implementations. Anyway, if this change is desired it would be a matter for a separate pull request.

File: crypto/rand/rand_win.c (hunk 24 above)
Comment (mattcaswell) [reply crypto/rand/rand_win.c]: Yes, I did mean that!

File: crypto/rand/rand_crng_test.c (hunk 23 above)
Comment (mattcaswell) [reply crypto/rand/rand_crng_test.c]: Effectively yes. I encountered a crash related to this code. The change is related to the `crngt_get_entropy` callback. That callback gets called from `rand_crng_ossl_ctx_new`. That function is called when we are initialising an OPENSSL_CTX with CRNG data for the first time. Unfortunately, in the original implementation, the default callback was itself trying to get CRNG data from the OPENSSL_CTX. But since we're still initialising it it went round in an an infinite recursive loop trying to initialise the OPENSSL_CTX until eventually it blew the stack. The solution to the problem was to change the callback to take the RAND_POOL to be used as a parameter instead of getting it from the OPENSSL_CTX.

File: crypto/rand/rand_win.c (hunk 24 above)
Comment (mspncp) [reply crypto/rand/rand_win.c]: Hmmm... but why? The code is marked deprecated-in-1.1.1 and not called anywhere.

```
msp@msppc:~/src/openssl$ git grep -E -nH 'RAND_(event|screen)' pr-9035
pr-9035:crypto/rand/rand_win.c:166:int RAND_event(UINT iMsg, WPARAM wParam, LPARAM lParam)
pr-9035:crypto/rand/rand_win.c:172:void RAND_screen(void)
pr-9035:doc/man3/RAND_add.pod:5:RAND_add, RAND_poll, RAND_seed, RAND_status, RAND_event, RAND_screen,
pr-9035:doc/man3/RAND_add.pod:25: int RAND_event(UINT iMsg, WPARAM wParam, LPARAM lParam);
pr-9035:doc/man3/RAND_add.pod:26: void RAND_screen(void);
pr-9035:doc/man3/RAND_add.pod:74:RAND_event() and RAND_screen() are equivalent to RAND_poll() and exist
pr-9035:doc/man3/RAND_add.pod:84:RAND_event() returns RAND_status().
pr-9035:doc/man3/RAND_add.pod:98:RAND_event() and RAND_screen() were deprecated in OpenSSL 1.1.0 and should
pr-9035:include/openssl/rand.h:68:DEPRECATEDIN_1_1_0(void RAND_screen(void))
pr-9035:include/openssl/rand.h:69:DEPRECATEDIN_1_1_0(int RAND_event(UINT, WPARAM, LPARAM))
pr-9035:util/libcrypto.num:1331:RAND_event                              1318    3_0_0   EXIST:_WIN32:FUNCTION:DEPRECATEDIN_1_1_0
pr-9035:util/libcrypto.num:1816:RAND_screen                             1804    3_0_0   EXIST:_WIN32:FUNCTION:DEPRECATEDIN_1_1_0
```

File: doc/internal/man3/ossl_prov_util_nid_to_name.pod [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+=head1 NAME
+
+ossl_prov_util_nid_to_name
+- provider utility functions
+
+=head1 SYNOPSIS
+
+ #include "internal/providercommon.h"
+
+ const char *ossl_prov_util_nid_to_name(int nid);
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply doc/internal/man3/ossl_prov_util_nid_to_name.pod]: @levitte has been championing the documentation of the "internal" API, i.e. stuff that ends up in our internal header files. There's quite a few man pages there already. This particular function is purely internal. There is no intention for it to be used by third party providers.

File: crypto/hmac/hm_meth.c [hunk 26]
```diff
@@ -152,13 +152,23 @@ static int hmac_ctrl_str(EVP_MAC_IMPL *hctx, const char *type,
 {
     if (!value)
         return 0;
+#ifndef FIPS_MODE
+    /*
+     * We don't have EVP_get_digestbyname() in FIPS_MODE. That function returns
+     * an EVP_MD without an associated provider implementation (i.e. it is
+     * using "implict fetch"). We could replace it with an "explicit" fetch
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply crypto/hmac/hm_meth.c]: Fixed.

File: crypto/rand/rand_win.c (hunk 24 above)
Comment (mspncp) [reply crypto/rand/rand_win.c]: Is it because it is still referenced in libcrypto.num?

```
pr-9035:util/libcrypto.num:1331:RAND_event                              1318    3_0_0   EXIST:_WIN32:FUNCTION:DEPRECATEDIN_1_1_0
pr-9035:util/libcrypto.num:1816:RAND_screen                             1804    3_0_0   EXIST:_WIN32:FUNCTION:DEPRECATEDIN_1_1_0
```

Comment (mattcaswell) [reply crypto/rand/rand_win.c]: Sorry, I wasn't clear. What *you* wrote is what I meant! I've fixed it now.

**Comment by mattcaswell** [comment]
Fix up commit pushed addressing the feedback.

File: crypto/rand/rand_win.c (hunk 24 above)
Comment (mspncp) [reply crypto/rand/rand_win.c]: But the FIPS provider does not reuse libcrypto.num, right?

Comment (mspncp) [reply crypto/rand/rand_win.c]: > Sorry, I wasn't clear. What _you_ wrote is what I meant! I've fixed it now.

Ok, now the world is in order again :-)

Comment (mattcaswell) [reply crypto/rand/rand_win.c]: > But the FIPS provider does not reuse libcrypto.num, right?

Right. It has its own num file which exports exactly one function:

https://github.com/openssl/openssl/blob/master/util/providers.num

**Comment by mattcaswell** [comment]
>  I'll do a live examination in the debugger this evening.

Did you manage to do that?

**Comment by mspncp** [comment]
Yes I did, and I have some thoughts to share, which I did not find the time to write down yet. Also, I might have some questions for you. But it's nothing that would prevent this pull request from being merged, I haven't found any issues (yet). I'd still be more comfortable if  @levitte would approve the provider related (non-drbg) parts of this pull request, but if you think my approval is sufficient, feel free to go ahead and merge it.

File: crypto/hmac/hm_meth.c (hunk 15 above)
Comment (mspncp) [reply crypto/hmac/hm_meth.c]: Has this request been resolved?

Comment (mattcaswell) [reply crypto/hmac/hm_meth.c]: I think this isn't so much a request to do anything on this PR...rather a side note.

**Comment by mattcaswell** [comment]
> I'd still be more comfortable if @levitte would approve the provider related (non-drbg) parts of this pull request, but if you think my approval is sufficient, feel free to go ahead and merge it.

I think @levitte is away at the moment with only occasional network access. I'll leave this open a little bit longer to see if he comments, but otherwise I will merge this later today.

**Comment by mattcaswell** [comment]
Pushed! Thanks!
